### ISSUE TITLE
Method to retrieve all event sources

### ISSUE THREAD

**Issue by arshaw** [fullcalendar/fullcalendar#2433]
Originally reported on Google Code with ID 2168
```
This has been implemented here, via @gmoore:
https://github.com/arshaw/fullcalendar/pull/105

Probably relates to Issue 678
```

Reported by `adamrshaw` on 2014-06-05 22:17:17

Imported with **2** stars.

**Comment by caseyjhol** [comment]
Do we want to implement this similar to #105, simply returning all event sources? Or, do we want to implement this similar to clientEvents, where you can pass through a filter or ID as well?

I realize event sources don't currently officially support having IDs, but if we're going to support refetching specific event sources, allowing them to have IDs might make things more convenient convenient (and wouldn't require any extra code as far as FullCalendar is concerned). Then, we can make the getEventSources API exactly like clientEvents (which would get my vote).

**Comment by arshaw** [comment]
i'd rather have it be more like #105 and only return an array. (i'd actually have it return a *copy* of the internal array to avoid mutations by the caller. learned this the hard way in the past).

I'm hesitant to add an ID system into event sources, and hesitant to add grepping functionality, as I'd prefer that to be the responsibility of the caller. My only argument: to avoid complexity. hopefully the caller can identify their own sources from the array result. no stopping them from putting their own ID in there.

FWIW, i regret having a `clientEvents` method that returns both a single-event and an array of events, depending on the input argument. I plan on breaking out into two methods: `getClientEvent(id)` and `getClientEvents()`

**Comment by arshaw** [comment]
@caseyjhol if you want to make a `getEventSource(id)` method that accepts a single "event source ID", i approve of that. event source IDs on the given source definition objects should be optional of course. plz make separate PR if u do so, thx

**Comment by caseyjhol** [comment]
Ok - I should be able to get those submitted this week.

**Comment by caseyjhol** [comment]
The first item in the `sources` array is always `{ events: [] }`. Since I'm planning on using slice anyway when returning the sources (to return a copy), I'm wondering if we should just slice out that first item? Or is there some potential usage for it that I'm missing?

**Comment by arshaw** [comment]
that initial empty event source is meant for sticky events from [renderEvent](http://fullcalendar.io/docs/event_rendering/renderEvent/).

i would NOT include that event source in the result for `getEventSources`.

ya, slicing out the first item sounds great. can u make sure to write a comment about what's going on?

**Comment by caseyjhol** [comment]
If I were to add a `getEventSource(id)` method, would you want it implemented similar to `getResourceById` in Scheduler? This would probably involve creating a sourcesById object that only tracks sources that actually have an ID (or we could assign an internal ID to the sources without IDs based on something like the index so the sourcesById object would contain all event sources). Or, it could just be a simple $.grep.

**Comment by arshaw** [comment]
simple grep is good. instead of the hassle of maintaining a hash. won't be that many resources anyway.

**Comment by arshaw** [comment]
[Code complete](https://github.com/fullcalendar/fullcalendar/commits/v2.8.0-dev) and slated for v2.8.0 ([wip docs](https://github.com/fullcalendar/fullcalendar/commits/docs-2.0)). Will notify when released.

**Comment by arshaw** [comment]
Released in [v2.8.0](https://github.com/fullcalendar/fullcalendar/releases/tag/v2.8.0).

http://fullcalendar.io/docs/event_data/getEventSources/
